concordium-node/src/bin/cli.rs

```diff
@@ -485,36 +488,105 @@ async fn maybe_do_out_of_band_catchup(
+        info!("Starting out of band catch-up");
+        let genesis_block_hashes = regenesis_arc.blocks.read().unwrap().clone();
+        if let Err(e) =
+            import_missing_blocks(consensus, &genesis_block_hashes, download_url, data_dir_path)
+                .await
+        {
+            if import_stopped.load(atomic::Ordering::Acquire) {
+                info!("Out of band catchup stopped.");
+            } else {
+                info!("Could not complete out of band catch-up due to: {:#}.", e);
             }
+        } else {
+            info!("Completed out of band catch-up from {}.", download_url)
         }
     }
 }
 
-async fn download_catchup_files(
-    download_url: url::Url,
+// An index entry for a chunk of blocks. Its format must correspond to one
+// produced by `database-exporter`.
+#[derive(serde::Deserialize)]
+struct BlockChunkData {
+    // exported chunk of blocks' filename
+    filename:           String,
+    // genesis block index from which relative heights of blocks in the chunk are counted
+    genesis_index:      usize,
+    // relative height of the oldest block stored in the chunk
+    first_block_height: u64,
+    // relative height of the newest block stored in the chunk
+    last_block_height:  u64,
+}
+
+async fn import_missing_blocks(
+    consensus: &ConsensusContainer,
+    genesis_block_hashes: &[HashBytes],
+    index_url: &url::Url,
     data_dir_path: &Path,
-) -> anyhow::Result<TempPath> {
+) -> anyhow::Result<()> {
+    let current_genesis_index = genesis_block_hashes.len() - 1;
+    let last_finalized_block_height = consensus.get_last_finalized_block_height();
+
+    info!("Current genesis index: {}", current_genesis_index);
+    info!("Local last finalized block height: {}", last_finalized_block_height);
+
+    let comments_stream = reqwest::get(index_url.clone())
+        .await?
+        .bytes_stream()
+        .map_err(|e| std::io::Error::new(std::io::ErrorKind::Other, e));
+    let mut comments_lines = tokio_util::io::StreamReader::new(comments_stream).lines();
+    while let Some(line) = comments_lines.next_line().await? {
+        if line.starts_with("# genesis hash ") {
+            let index_genesis_block_hash = line.strip_prefix("# genesis hash ").unwrap();
+            let genesis_hash = genesis_block_hashes[0].to_string();
+            if index_genesis_block_hash != genesis_hash {
+                return Err(anyhow::anyhow!(
+                    "The genesis block hash in the catchup index file {} does not match the \
+                     genesis block hash {} in the local tree state. Please verify that you chose \
+                     the catchup service for the correct chain.",
+                    index_genesis_block_hash,
+                    genesis_hash
+                ));
+            } else {
+                break;
+            }
+        }
+    }
+
+    let index_reader = reqwest::get(index_url.clone())
+        .await?
```

Comment:
I'm sorry, but I just cannot accept downloading the file twice. I understand that it is in practice not a lot of data, but it is also not free.

The only reason for doing this seems to be that the file format is not very good for not downloading the file twice.

It seems it would be easy to change the file format and define it as
- first line: hash of the genesis of the chain
- remaining lines, comma separated list as now. You could even include the headers if you wanted to as a first line here, but I don't think it is necessary.

That would allow you to parse the first line in a special way, and the remaining ones as CSV. Since the format is so simple you don't need a CSV parser, you can just parse it directly.

Reply:
A hand-written parser for a format even as simple as a custom dialect of CSV is going to put unnecessary mental strain the next person reading this code. I suggest we use the existing library.

Removing field headers from the file is, indeed, making the file more amenable to parsing in a single streaming. Done.